prov/udp/src/udpx_attr.c

```diff
@@ -93,6 +94,69 @@ struct fi_info udpx_info = {
 
 struct util_prov udpx_util_prov = {
 	.prov = &udpx_prov,
-	.info = &udpx_info,
-	.flags = 0,
+	.info = NULL,
+        .flags = 0,
 };
+
+
+static int match_interface(struct slist_entry *entry, const void *infop)
+{
+	struct ofi_addr_list_entry *addr_entry;
+	const struct fi_info* info = infop;
+
+	addr_entry = container_of(entry, struct ofi_addr_list_entry, entry);
+	return strcmp(addr_entry->net_name, info->fabric_attr->name) == 0 &&
+	       strcmp(addr_entry->ifa_name, info->domain_attr->name) == 0;
+}
+
+static void set_mtu_from_addr_list(struct fi_info* info,
+				   struct slist *addr_list)
+{
+	struct ofi_addr_list_entry *addr_entry;
+	struct slist_entry *entry;
+	int max_msg_size;
+
+	entry = slist_find_first_match(addr_list, match_interface, info);
+	if (entry) {
+		addr_entry = container_of(entry,
+					  struct ofi_addr_list_entry,
+					  entry);
+		max_msg_size = UDPX_MAX_MSG_SIZE(addr_entry->mtu);
+		if (max_msg_size > 0) {
+			info->tx_attr->inject_size = max_msg_size;
+			info->ep_attr->max_msg_size = max_msg_size;
+		}
+	} else {
+		FI_DBG(&udpx_prov, FI_LOG_CORE,
+		       "Failed to match interface (%s, %s) to "
+		       "address for MTU size\n",
+		       info->fabric_attr->name, info->domain_attr->name);
+	}
+}
+
+void udpx_util_prov_init(uint32_t version)
+{
+
+	struct slist addr_list;
+	struct fi_info* cur;
+	struct fi_info* info;
+
+        if (udpx_util_prov.info == NULL) {
+		udpx_util_prov.info = &udpx_info;
+		info = fi_allocinfo();
+		ofi_ip_getinfo(&udpx_util_prov, version, NULL, NULL, 0, NULL,
+			       &info);
```

Comment:
I think the way to handle this is to add a callback to ofi_ip_getinfo().  That function is already iterating over the interfaces.  The only part that's missing is the ability of the provider to update the fi_info based on the specific interface.  For most callers, the callback will be null.  For UDP, the callback will receive the fi_info and ofi_addr_list_entry, so that it can update the attributes based on the mtu.  The callback may need to be passed through to subfunctions.

Reply:
I will do that. It seems like a more extensive change to the code, but if you're looking over my shoulder, I'll give it a try.